### ISSUE TITLE
Feature list

### ISSUE THREAD

**Issue by thomashopf** [evedesignbio/evedesign#1]
## Technical features
- [x] Proper types and checking
- [x] Allow to instantiate from JSON specification of models, constraints, molecular systems etc

## Scientific general features
- [x] Flexible general specification of molecular system, including multimers (and later possible non-protein components)
- [x] What is best way to sample from models trained on experimental data?
- [ ] Allow specification of different states (e.g. conformations) and context of system? (hold for now)
- [ ] Map resources to different components / states (hold for now)

## MSA generation
- [ ] Automatic choice of database (viral/higher eukaryote/other)
- [x] Redundancy reduction server-side?

## Gibbs sampler
- [x] Flexible specification of iteration order
- [x] Efficient constraints (than can cover more than just the sampled positions for context)
- [x] Add callback for sampling status (steps sampled)
- [ ] Variable-length sampling via general MCMC/Metropolis? (make this a separate sampler)
- [ ] Support greedy sampling (no, make this a separate sampler)

## Autoregressive sampler
- [x] Would this be able to handle constraints as well? Eg. PSSM? (difficult unless we hack into each implementation - ideally would have a callback for sampling step allowing to combine model logits with other information)

## Models 
- [x] Generic API covering all models
- [x] Exposes minimum needed to sampler (closure-like around molecular system)
- [x] Add method to estimate resource requirements 
- [x] Allow to check if model can handle a certain problem specification
- [x] Track sequences and structures (and/or any other metadata?)
- [x] Methods: fit, score, embed, generate, serialize, load, check, score_singles
- [x] Support serialization and deserialization (also nested)
- [x] Add callback for training/inference status

## Working hypothesis on labelled data integration / model nesting
- [x] Supply other models via constructor; can be pre-built for reusing, or raw (will be checked upon execution)?
- [x] Can we supply raw embeddings? Does this even make sense (probably not)?

**Comment by thomashopf** [comment]
all relevant points added for now
